Add support for metadata component

artem-smotrakov opened the pull request:
Part of #6

madpah commented:
Thanks for the PR @artem-smotrakov. We're in the process of a notable update at present to support the forthcoming CycloneDX 1.4 schema - so please forgive us if we wait for that work to stabilise before we look at merging your PR.

Cheers!

jkowalleck commented:
> Thanks for the PR @artem-smotrakov. We're in the process of a notable update at present to support the forthcoming CycloneDX 1.4 schema - so please forgive us if we wait for that work to stabilise before we look at merging your PR.
> 
> Cheers!


enabled CI to run CT/QA.

@artem-smotrakov could you check that all tests pass?
see https://github.com/CycloneDX/cyclonedx-python-lib/blob/main/CONTRIBUTING.md for more.

artem-smotrakov commented:
Hmm, I read the contributing guidelines and ran tests and `poetry run autopep8 --in-place -r .` before opening the PR. They passed. Let me check again.

artem-smotrakov commented:
I've fixed several formatting and typing issues. `poetry run tox` passed. Please re-run the CI/CD jobs.

madpah commented:
Hi @artem-smotrakov - we've just merged changes for CycloneDX 1.4 - are you willing to rebase this branch from main?

Thanks, Paul

artem-smotrakov commented:
@madpah I've rebased it -- please rerun CI/CD and have a look at the updates.

I am not sure about this one:

https://github.com/CycloneDX/cyclonedx-python-lib/pull/118/files#diff-9c84a99c5332704dd6a90a4defadfd9df437336f07b79b91033e0a96bbd8fbf1R95

I've noticed that if a BOM doesn't have components, `_specialise_output_for_schema_version()` fails with a key error. I've updated it to always add `components` element. That is how XML encoder behaves.

madpah commented on cyclonedx/model/bom.py:
```diff
@@ -80,6 +82,30 @@ def timestamp(self) -> datetime:
+    @component.setter
+    def component(self, component: Component) -> None:
+        """
+        Set a Component for this Bom Metadata.
```
@artem-smotrakov - I've tried to mirror the descriptions provided in the Specification here - so ideally this doc would read:

> The (optional) component that the BOM describes.

madpah commented on cyclonedx/output/xml.py:
```diff
@@ -113,6 +113,9 @@ def _add_metadata_element(self) -> None:
             for tool in bom_metadata.tools:
                 self._add_tool(parent_element=tools_e, tool=tool)
 
+        if bom_metadata.component is not None:
```
You don't need the `is not None` - `if bom_metadata.component` suffices

madpah reviewed (changes requested):
@artem-smotrakov - thanks again for the PR.

It's interesting you have pointed out that `components` is output as an empty element in some cases - ideally this should not be output if there are no Components - but we are happy to resolve this as a separate bug later.

artem-smotrakov commented:
@madpah I've addressed your comments -- thanks for the review!
